File: Data Structures/implementation of data structures/linked_list/doubly_ll.cpp
#include <bits/stdc++.h>
using namespace std;


class Node{
	public:
		Node *next;
		Node *pre;
		int data;
		Node(int d){
			data = d;
		}
};

class doubly_ll{
	public:
		Node *head;
		Node *tail;

	doubly_ll(){
		head = tail = NULL;
	}

	void display(){
		Node *temp = head;
		while(temp != NULL){
			cout << temp->data << " -> " ;
			temp = temp->next;
		}
		cout << "null" << endl;
	}

	void insert_head(int d){
		Node *node = new Node(d);
		if(head == NULL){
			head = tail = node;
			return;
		}

		node->next = head;
		head->pre = node;
		head = node;
	}	

	void insert_tail(int d){
		Node *node = new Node(d);
		if(tail == NULL){
			head = tail = NULL;
			return;
		}

		tail->next = node;
		node->pre = tail;
		tail = node;
	}

	bool isEmpty(){
		return head == NULL || tail == NULL;
	}

	void delete_head(){
		if(!isEmpty()){
			if(head->next == NULL){
				head = tail = NULL;
				return;
			}
			Node *temp = head;
			temp->next->pre = NULL;
			head = temp->next;
			temp->next = NULL;
		}else{
			cout << "empty" << endl;
		}
	}
	void delete_tail(){
		if(!isEmpty()){
			if(head->next == NULL){
				head = tail = NULL;
				return;
			}
			Node *temp = tail;
			temp->pre->next = NULL;
			tail = temp->pre;
			temp->pre = NULL;
			
		}else{
			cout << "empyt" << endl;
		}
	}

		void insert_at_key(int d, int key){
			Node *node = new Node(d);
			Node *temp = head;
			for(int i = 0; i < key - 1 && temp->next != NULL; i++){
				temp = temp -> next;
			}

			if(key == 0){
				insert_head(d);
				return;
			}
			if(temp == tail){  // if index is greate than size
				insert_tail(d);
				return;
			}
			node->next = temp->next;
			temp->next->pre = node;
			temp->next = node;
			node->pre = temp;
		}

		void delete_at_key(int key){
			if(!isEmpty()){
				Node *current = head;
				Node *previous = NULL;
				for(int i = 0; i < key && current->next != NULL; i++){
					previous = current;
					current = current -> next;
				}
				if(current == head){
					delete_head();
					return;
				}
				if(current == tail){
					delete_tail();
					return;
				}
				current->pre = NULL;
				previous->next = current->next;
				previous->next->pre = previous;
				current->next = NULL;
			}else{
				cout << "empty " << endl;
			}
		}
};

int main(){
	doubly_ll ll;
	int c;
	do{
        cout << "enter your choice \n1.insert head \n2. insert tail \n3. delete head \n4. delete tail \n5. insert_at_key \n6. delete_at_key" << endl;
        int choice;
        cin >> choice;
        switch (choice){
            case 1 :
               cout << "enter the data : ";
                int a ;
                cin >> a;
                ll.insert_head(a);
                ll.display();
                break;
            case 2 :
               	cout << "enter the data : ";
                int b;
                cin >> b;
                ll.insert_tail(b);
                ll.display();
                break;
            case 3 :
            	ll.delete_head();
            	ll.display();
            	break;
        	case 4:
        		ll.delete_tail();
        		ll.display();
        		break;
            case 5:
            	cout << "enter the index ";
            	int key;
            	cin >> key;
               	cout << "enter the data : ";
                int c;
                cin >> c;
                ll.insert_at_key(c, key);
                ll.display();
                break;
            case 6:
            	cout << "enter the index : " ;
            	int k;
            	cin >> k;
            	ll.delete_at_key(k);
            	ll.display();
            	break;	    

            default:
                cout << "invalid input";
        }
        cout << "\ndo you wanna continue 1.yes 2.no ";
        cin >> c;
	}while (c == 1);
	return 0;
}

File: Data Structures/implementation of data structures/linked_list/doublyLL.java
package linked_list;

import  java.util.*;

public class doublyLL {
    public  class Node {
        public Node next;
        public Node prev;
        public int data;
        public Node(int d) {
            data = d;
//            next = prev = null;
        }
    }
    public Node head; // by default value is null
    public Node tail;

    public doublyLL(){
        head = tail = null;
    }

    public void display(){
        Node temp = head;
        while ( temp != null) {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        }
        System.out.println("null");
    }
    public boolean isEmpty(){
        return head == null;
    }

    public void insertHead(int d){
        Node node = new Node(d);
        if(head == null){
            head = node;
            tail = node;
        }
        else{
            node.next = head;
            head.prev = node;
            head = node;
        }
    }
    public void insertTail(int d){
        Node node = new Node(d);
        if(tail == null) {
           tail = node;
           head = node;
        }
        else {
            tail.next = node;
            node.prev = tail;
            tail = node;
        }
    }

    public void deleteFirst(){
        Node temp = head;
        if(head.next == null){
            head = tail = null;
        }
        else{
            temp.next.prev = null;
            head = temp.next;
            temp.next = null;
        }
    }
    public void deleteLast(){
        Node temp = tail;
        if(tail.prev == null){
            head = tail = null;
        }
        else{
            temp.prev.next = null;
            tail = tail.prev;
            temp.prev = null;
        }
    }
    void insetATkey(int d, int key){
        Node node = new Node(d);
        Node temp = head;
        for(int i = 0; i < key - 1 && temp.next != null; i++){
            temp = temp.next;
        }

        if(key == 0){
            insertHead(d);
            return;
        }
        if(temp == tail){  // if index is greater than size
            insertTail(d);
            return;
        }
        node.next = temp.next;
        temp.next.prev = node;
        temp.next = node;
        node.prev = temp;
        display();
//        delete(node);
    }

    void deleteATkey(int key){
        if(!isEmpty()){
            Node current = head;
            Node previous = null;
            for(int i = 0; i < key && current.next != null; i++){
                previous = current;
                current = current.next;
            }
            if(current == head){
                deleteFirst();
                return;
            }
            if(current == tail){
                deleteLast();
                return;
            }
            current.prev = null;
            previous.next = current.next;
            previous.next.prev = previous;
            current.next = null;
        }else{
            System.out.println("empty");
        }
    }

    public int sizeofLL(){
        Node temp = head;
        int s = 0;
        while(temp!=null){
            temp = temp.next;
            s++;
        }
        return s;
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int c,key;

        doublyLL ll = new doublyLL();
        do{
            System.out.println("enter your choice \n1.insert first \n2.insert last \n3.delete first " +
                    "\n4.delete last \n5.insert at a key \n6.delete at a key ");
            int choice = sc.nextInt();
            switch (choice){
                case 1:
                    System.out.print("enter the data :");
                    int d = sc.nextInt();
                    ll.insertHead(d);
                    ll.display();
                    break;
                case 2:
                    System.out.print("enter the data :");
                    int a = sc.nextInt();
                    ll.insertTail(a);
                    ll.display();
                    break;
                case 3:
                    if(!ll.isEmpty()){
                        ll.deleteFirst();
                        ll.display();
                        break;
                    }
                    else{
                        System.out.println("empty list");
                        break;
                    }
                case 4:
                    if(!ll.isEmpty()){
                        ll.deleteLast();
                        ll.display();
                        break;
                    }
                    else{
                        System.out.println("empty list");
                        break;
                    }
                case 5:
                    int Size = ll.sizeofLL();
                    System.out.println("size = " + Size);
                    System.out.print("enter the index:");
                    key = sc.nextInt();
                    if(Size < key){
                        System.out.println("invalid key");
                        break;
                    }
                    System.out.print("enter the data :");
                    int data = sc.nextInt();
                    ll.insetATkey(data, key);
                    ll.display();
                    break;

                case 6:
                    int size = ll.sizeofLL();
                    System.out.println("size = " + size);
                    if(size == 0 ){
                        System.out.println("empty list");
                        break;
                    }
                    System.out.print("enter the index:");
                    key = sc.nextInt();
                    if(size <= key){
                        System.out.println("invalid key");
                        break;
                    }
                    ll.deleteATkey(key);
                    ll.display();
                    break;

                default:
                    System.out.println("invalid input");
            }
            System.out.println("\ndo you wanna continue 1.yes 2.no ");
            c = sc.nextInt();
        }while (c == 1);
    }
}
